Adds Transfer Wrapped Assets guide to Token Bridge section (#461)
wormhole-foundation/wormhole-docs: products/token-bridge/guides/transfer-wrapped-assets.md

            --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/attestToken.ts::107'
            ```

            This code does the following:
        
            - Gets the Token Bridge protocol for the source chain.
            - Defines the token to attest for registration on the destination chain and the payer to sign for the transaction.
            - Calls `createAttestation`, signs, and then sends the transaction.
            - Waits for the signed VAA confirming the attestation creation.
            - Sends the VAA to the destination chain to complete registration.
            - Polls for the wrapped token to be available on the destination chain before continuing the transfer process.

        3. Run the script with the following command:
            
            ```bash
            npx tsx attestToken.ts
            ```

            When the attestation and registration are complete, you will see terminal output similar to the following:

            --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/terminal-2.html'

        You can now go on to [initiate the transfer](#initiate-transfer-on-source-chain) on the source chain.

> ## Initiate Transfer on Source Chain

Reviewer: I think it might be nice to briefly describe the difference between automatic vs manual - it is sort of covered in the comments in the code but might be nice to provide this info explicitly and upfront.

Reviewer 2: I added a link to this page: https://wormhole.com/docs/products/token-bridge/concepts/transfer-flow/#automatic-vs-manual-transfers
Otherwise, the page will be too long.

Follow these steps to add the rest of the logic to initiate the token transfer on the source chain:

> 1. Open your `transfer.ts` file and add the following code:

Reviewer: Where in the `transfer.ts` file does this need to be added? Although might be obvious to add it to the end, let's make it super clear

>     ```typescript title="transfer.ts"

Reviewer: I still think tabs for automatic vs manual might be helpful here, because if going with manual there's some automatic variables that aren't necessary, and if going with automatic, there's extra unnecessary logic for completing the transfer

Reviewer: For manual transfers, maybe showcase the `isTransferCompleted` function at the end?

Author: I added the tabs and I like it! Attempted to separate these out to just what's needed for manual and auto on each tab but I haven't tested them yet.

    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer.ts:49:88'
    ```

    This code does the following:

>     - Defines the transfer as automatic or manual. To use automatic transfer, both the source and destination chain must have an existing `tokenBridgeRelayer` contract. You can check the list of [deployed `tokenBridgeRelayer` contracts](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/a48c9132015279ca6a2d3e9c238a54502b16fc7e/core/base/src/constants/contracts/tokenBridgeRelayer.ts){target=\_blank} in the Wormhole SDK repo to see if your desired chains are supported.

Reviewer: ```suggestion
    - Defines the transfer as automatic or manual. For automatic transfers, both the source and destination chain must have an existing `TokenBridgeRelayer` contract, which listens for and completes transfers on your behalf. You can check the list of [deployed `TokenBridgeRelayer` contracts](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/a48c9132015279ca6a2d3e9c238a54502b16fc7e/core/base/src/constants/contracts/tokenBridgeRelayer.ts){target=\_blank} in the Wormhole SDK repo to see if your desired chains are supported.
```

I think it's implied, but can add a tiny bit more context here to make it super clear

    - Sets an optional amount for native gas drop-off. This option allows you to send a small amount of the destination chain's native token for gas fees. Native gas drop-off is currently only supported for automatic transfers.

>     - Builds the transfer object, initiates the transfer, signs and sends the transaction.

Reviewer: ```suggestion
    - Builds the transfer object, initiates the transfer, signs the transaction, and sends it.
```

    - If the transfer is automatic, the flow ends. Otherwise, the script waits for the signed VAA confirming the transaction on the source chain. The signed VAA is then submitted to the destination chain to claim the tokens and complete the manual transfer.

> 

Reviewer: What if a user wants to send a payload? What would that look like? I think this is the time to go over all of the different options available

Reviewer: Also, I've started to think about automatic transfers a bit more, we don't really mention anything about fees, what are the fees, how do they paid for, how much can be swapped for native tokens. Maybe we can briefly mention some of these things here, maybe other items would be better addressed in a separate guide (i.e., reference docs on maxSwapAmount function could have an example on how to use that function to get the max amount)

Reviewer 2: I added a link to the automatic transfer flow where there is a mention about how the gas drop-off works. If we want to mention more details about gas drop-off, maxSwapAmount, etc., they should be on that page. Otherwise, this guide could get too long with definitions.

---
title: Transfer Wrapped Assets
description: Follow this guide to use Token Bridge to transfer wrapped assets. Includes automatic and manual flows, token attestation, VAA fetching, and manual redemption.
categories: Token-Bridge, Transfers, Typescript-SDK
---

# Transfer Wrapped Assets

## Introduction

> This guide demonstrates the transfer of wrapped assets using the Token Bridge protocol via the TypeScript SDK. This example will transfer an arbitrary ERC-20 token from Moonbase Alpha to Solana but can be adapted for any supported chains. View this list of chains with [deployed Token Bridge contracts](/products/reference/contract-addresses/#token-bridge){target=\_blank} to verify if your desired source and destination chains are supported.

Reviewer 2: I would link it to the Typescript SDK Github repo

> This guide demonstrates the transfer of wrapped assets using the Token Bridge protocol via the TypeScript SDK. This example will transfer an arbitrary ERC-20 token from Moonbase Alpha to Solana but can be adapted for any supported chains. View this list of chains with [deployed Token Bridge contracts](/products/reference/contract-addresses/#token-bridge){target=\_blank} to verify if your desired source and destination chains are supported.

Reviewer 2: ```suggestion
This guide demonstrates the transfer of wrapped assets using the Token Bridge protocol via the TypeScript SDK. This example will transfer an arbitrary ERC-20 token from Moonbase Alpha to Solana but can be adapted for any supported chains. View this list of chains with [deployed Token Bridge contracts](/docs/products/reference/contract-addresses/#token-bridge){target=\_blank} to verify if your desired source and destination chains are supported.
```

## Verify Token Registration (Attestation)

Tokens must be registered on the destination chain before they can be bridged. This process involves submitting an attestation with the native token metadata to the destination chain, which enables the destination chain's Token Bridge contract to create a corresponding wrapped version with the same attributes as the native token.

Registration via attestation is only required the first time a given token is sent to that specific destination chain. Follow these steps to check the registration status of a token:

1. Create a new file named `transfer.ts`:
   ```bash
   touch transfer.ts
   ```

2. Open your `transfer.ts` file and add the following code:
    ```typescript title="transfer.ts"
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:1:47'
    // Token attestation and registration flow here if needed
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:127:127'

>     --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:171:174'

Reviewer 2: The snippet is missing a "}" at the end for it to work

    ```

    This code does the following:

    - Initializes a `wormhole` instance and defines the source and destination chains
    - Imports the signer and decimal functions from `helpers.ts`
    - Identifies the token and amount to transfer
    - Checks to see if a wrapped version of the ERC-20 token to transfer exists on the destination chain

3. Run the script using the following command:

    ```bash
    npx tsx transfer.ts
    ```

>     If the token is registered on the destination chain, the address of the existing wrapped asset is returned, and you can continue to [initiate the transfer](#initiate-transfer-on-source-chain) on the source chain. If the token is not registered, you will see a message similar to the following advising the attestation flow will run:

Reviewer 2: When I run it, the token address is not returned. I know the token is already wrapped because I used it in the past, and it doesn't throw an error.

            - Waits for the signed VAA confirming the attestation creation
            - Sends the VAA to the destination chain to complete registration
            - Polls for the wrapped token to be available on the destination chain before continuing the transfer process

        3. Run the script with the following command:
            
            ```bash
            npx tsx transfer.ts
            ```

            When the attestation and registration are complete, you will see terminal output similar to the following:

            --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/terminal02.html'

        You can now go on to [initiate the transfer](#initiate-transfer-on-source-chain) on the source chain.

## Initiate Transfer on Source Chain

Follow these steps to add the rest of the logic to initiate the token transfer on the source chain:

1. Open your `transfer.ts` file and add the following code:

    ```typescript title="transfer.ts"
    // Remainder of transfer code 
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:129:174'
         
    ```

    This code does the following:

>     - Defines the transfer as automatic or manual. To use automatic transfer, both the source and destination chain must have an existing `tokenBridgeRelayer` contract. You can check the list of [deployed `tokenBridgeRelayer` contracts](https://github.com/wormhole-foundation/wormhole-sdk-ts/core/base/src/constants/contracts/tokenBridgeRelayer.ts) in the Wormhole SDK repo to see if your desired chains are supported

Reviewer 2: The link to the contracts is returning a 404

    - Sets an optional amount for native gas drop-off. This option allows you to send a small amount of the destination chain's native token for gas fees. Native gas drop-off is currently only supported for automatic transfers
    - Builds the transfer object, initiates the transfer, signs and sends the transaction
    - If the transfer is automatic, the flow ends. Otherwise, the script waits for the signed VAA confirming the transaction on the source chain. The signed VAA is then submitted to the destination chain to claim the tokens and complete the manual transfer

2. Run the script with the following command:
    ```bash

>     npx tsx transfer.ts

Reviewer: This comment still applies

>     npx tsx transfer.ts

Author: The polling logic we discussed was for the token attestation script to poll for the wrapped version until it became available and then return the address for it.

>     npx tsx transfer.ts

Reviewer 2: @eshaben this happens when the set timeout is too low. Depending on the blockchain, some take more time than others. I will increase the timeout.

---
title: Transfer Wrapped Assets
description: Follow this guide to use Token Bridge to transfer wrapped assets. Includes automatic and manual flows, token attestation, VAA fetching, and manual redemption.

> categories: Token-Bridge, Transfers, Typescript-SDK

Reviewer 3: ```suggestion
categories: Token Bridge, Transfer, Typescript SDK
```

Registration via attestation is only required the first time a given token is sent to that specific destination chain. Follow these steps to check the registration status of a token:

1. Create a new file named `transfer.ts`:
   ```bash
   touch transfer.ts
   ```

2. Open your `transfer.ts` file and add the following code:
    ```typescript title="transfer.ts"
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer.ts::43'
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer.ts:80:85'
    ```

    This code does the following:

    - Initializes a `wormhole` instance and defines the source and destination chains.
    - Imports the signer and decimal functions from `helpers.ts`.
    - Identifies the token and amount to transfer.
    - Checks to see if a wrapped version of the ERC-20 token to transfer exists on the destination chain.

3. Run the script using the following command:

    ```bash

>     npx tsx transfer.ts

Reviewer 3: why are we running the script now if I am missing the instructions for Insert Initiate Transfer on Source Chain code that follow later

Reviewer 2: That's to verify if the token was attested on the destination chain. If it's not, it will let you know before proceeding.

Reviewer 3: then it needs some closing } because it won't work like this if they just copy paste it

Registration via attestation is only required the first time a given token is sent to that specific destination chain. Your transfer script should include a check for an existing wrapped version of the token on your destination chain. Follow these steps to check the registration status of a token:

1. Inside your `src` directory, create a new file named `transfer.ts`:
   ```bash
   touch transfer.ts
   ```

2. Open your `transfer.ts` file and add the following code:
    ```typescript title="transfer.ts"
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:1:91'
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:139:142'
    ```

    This code does the following:

    - Initializes a `wormhole` instance and defines the source and destination chains
    - Imports the signer and wallet functions from `helpers.ts`
    - Identifies the token to transfer and verifies the token balance in the source wallet
    - Gets the `TokenBridge` protocol client for the source chain
    - Checks to see if a wrapped version of the ERC-20 token to transfer exists on the destination chain

> 3.  Run the script using the following command:

Reviewer 4: ```suggestion
3. Run the script using the following command:
```

    ```bash
    npx tsx transfer.ts
    ```

    If the token is registered on the destination chain, the address of the existing wrapped asset is returned, and you can continue to [initiate the transfer](#initiate-transfer-on-source-chain) on the source chain. If the token is not registered, you will see a message similar to the following advising attestation is required:

    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/terminal01.html'

>     ??? example "Need to register a token?"

Reviewer: @martin0995 - please double-check and if it looks good to you resolve convo

        3. Run the script with the following command:
            
            ```bash
            npx tsx attestToken.ts
            ```

            When the attestation and registration are complete, you will see terminal output similar to the following:

            --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/terminal02.html'

        You can now go on to [initiate the transfer](#initiate-transfer-on-source-chain) on the source chain.

## Initiate Transfer on Source Chain

Follow these steps to add the rest of the logic to initiate the token transfer on the source chain:

1. Open your `transfer.ts` file and replace the commented line "// Additional transfer code" with the following code:

    ```typescript title="transfer.ts"
    --8<-- 'code/products/token-bridge/guides/transfer-wrapped-assets/transfer01.ts:92:138'
    ```

    This code does the following:

    - Uses the supplied [Token Bridge contract address](https://wormhole.com/docs/build/reference/contract-addresses/#token-bridge){target=\_blank} to approve spending the ERC-20 token in the amount you want to transfer
    - Calls the `transfer()` method to initiate the transfer on the source chain

>     - Watches for the transaction, parses the transaction ID to read the Wormhole message, and waits for the Guardians to sign the VAA verifying the transaction

Reviewer 4: ```suggestion
    - Watches for the transaction, parses the transaction ID to read the Wormhole message and waits for the Guardians to sign the VAA verifying the transaction
```
